### PR TITLE
fix(func): raise explicit error for grad/vjp/jvp in inference_mode

### LINKED ISSUE #177318: torch.func transforms silently return zeros under inference_mode
## 🐛 Describe the bug

  `torch.func.{grad, vjp, jvp}` silently produce incorrect results (zeros)
  when called under `torch.inference_mode()`. No error is raised.

  ```python
  import torch
  from torch.func import jvp

  x = torch.randn(3)
  with torch.inference_mode():
      out, tangent = jvp(lambda x: x**2, (x,), (torch.ones(3),))
      print(tangent)  # tensor([0., 0., 0.]) — should be 2*x
```
##  Root cause

  Functorch transforms set up their own autograd context on entry:
  enable_grad() for grad/vjp, _set_fwd_grad_enabled(True) for jvp
  (see NOTE [grad and vjp interaction with no_grad]). However,
  inference_mode also excludes autograd dispatch keys from the TLS
  dispatch key set, and the mode-enabling calls don't undo that. So
  autograd dispatch keys remain excluded and differentiation silently
  produces zeros.

  The transforms already correctly handle no_grad — this is the same
  class of issue with a different TLS mechanism.

##  Expected behavior

  Either produce correct results (by also disabling inference_mode on
  transform entry, matching the existing no_grad handling) or raise
  an error.

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93 @Chillee @samdow @kshitij12345

### PR DESCRIPTION
**Pull request by Lillian68**
## Summary
Prevent `torch.func.{grad, vjp, jvp}` from silently returning incorrect zero results when called under `torch.inference_mode()` by raising an explicit error.

### Background
Currently, gradient transforms (grad/vjp/jvp) in `torch.func` fail silently (return zeros) when run in `inference_mode()`:
- Root cause: `inference_mode()` excludes autograd dispatch keys from TLS, and functorch's internal `enable_grad()` does not undo this exclusion
- This behavior is unsafe (hard to debug) and inconsistent with PyTorch's "fail fast" philosophy

### Solution
Add a check at the C++ entry of `grad`/`vjp`/`jvp` transforms:
1. Detect if the current context is `inference_mode()` (via `c10::InferenceMode::is_enabled()`)
2. Raise a `TORCH_CHECK` with clear error message if true (instead of silent failure)
3. Error message example: 
   `torch.func.{grad, vjp, jvp} are not supported in torch.inference_mode(). 
     Please exit inference_mode before calling these transforms, as autograd is required for differentiation.`

### Rationale
- `inference_mode()` is explicitly designed for non-differentiable inference workloads — allowing gradient transforms to run here is semantically inconsistent
- Explicit error > silent failure: aligns with PyTorch's design principles and avoids hard-to-debug user issues
- Consistent with existing handling of `no_grad()` (functorch already handles `no_grad()` correctly, this extends the same "fail fast" logic to `inference_mode()`)

### Unit Test
- Unit test for `grad`/`vjp`/`jvp` under `inference_mode()`: verify that an error is raised (not zeros returned) and whether results are correct without inference_mode()
- Test location: `test/functorch/test_eager_transforms.py`
- add new test function named `test_inference_mode_gradient_transforms` in class `TestGradTransform`

### User Test
result

## Related Issues
Fixes #177318

### CODE REVIEW

File: torch/_functorch/eager_transforms.py [hunk 1]
```diff
@@ -365,6 +365,20 @@ def jvp_increment_nesting() -> Generator[int, None, None]:
         exit_jvp_nesting()
 
 
+def _check_inference_mode_enabled():
+    # Check for inference_mode: it disables autograd dispatch keys, leading to silent zero gradients
+    # Skip during Dynamo compilation: avoid graph breaks from tracing torch.is_inference_mode_enabled()
+    if torch._dynamo.is_compiling():
```
Comment (soulitzer) [inline torch/_functorch/eager_transforms.py]: What happens in compile? We still return zeros?

Comment (Lillian68) [reply torch/_functorch/eager_transforms.py]: First of all, the reason I added the torch._dynamo.is_compiling check is that an error occurred in the TestCompileTransformsCPU.test_grad_deprecated_api_cpu test case, as follows:
```
@torch._dynamo.config.patch(suppress_errors=False)
    def test_grad_deprecated_api(self, device):
        x = torch.randn((), device=device)
        y = torch.randn((), device=device)

        def wrapper_fn(x, y):
            return functorch.grad(torch.mul)(x, y)

        actual = wrapper_fn(x, y)
        expected = torch.compile(wrapper_fn, backend="eager", fullgraph=True)(x, y)
        torch.compile(wrapper_fn, backend="eager", fullgraph=True)
        self.assertEqual(actual, expected)

        def wrapper_fn(x, y):
            return functorch.grad(torch.mul, argnums=(0, 1))(x, y)

        actual = wrapper_fn(x, y)
        expected = torch.compile(wrapper_fn, backend="eager", fullgraph=True)(x, y)
        self.assertEqual(actual, expected)
```
截屏2026-03-17 21 05 15

Regarding the behavior during compilation, I have conducted a detailed test with the results below:
```
import torch
from torch.func import jvp


# test case 1: use jvp in torch.inference_mode
x = torch.randn(3)
with torch.inference_mode():
    try:
        out, tangent = jvp(lambda x: x**2, (x,), (torch.ones(3),))
        print("1-use jvp in torch.inference_mode → tangent:", tangent)
    except Exception as e:
        print("1-use jvp in torch.inference_mode → tangent:", e)


# test case 2: use jvp in torch.compile with fullgraph=True
@torch.compile(backend="eager", fullgraph=True)
def compiled_jvp(x):
    out, tangent = jvp(lambda x: x**2, (x,), (torch.ones(3),))
    return tangent

try:
    tangent_compiled = compiled_jvp(x)
    print("2-use jvp in torch.compile with fullgraph=True → tangent:", tangent_compiled)
except Exception as e:
    print("2-use jvp in torch.compile with fullgraph=True → tangent:", e)


# test case 3: use jvp in torch.compile with fullgraph=False
@torch.compile(backend="eager", fullgraph=False)
def compiled_jvp(x):
    out, tangent = jvp(lambda x: x**2, (x,), (torch.ones(3),))
    return tangent

try:
    tangent_compiled = compiled_jvp(x)
    print("3-use jvp in torch.compile with fullgraph=False → tangent:", tangent_compiled)
except Exception as e:
    print("3-use jvp in torch.compile with fullgraph=False → tangent:", e)


# test case 4: use jvp in torch.compile with fullgraph=True and inference_mode
@torch.compile(backend="eager", fullgraph=True)
def compiled_jvp(x):
    with torch.inference_mode():
        out, tangent = jvp(lambda x: x**2, (x,), (torch.ones(3),))
        return tangent
    
try:
    tangent_compiled = compiled_jvp(x)
    print("4-use jvp in torch.compile with fullgraph=True and inference_mode → tangent:", tangent_compiled)
except Exception as e:
    print("4-use jvp in torch.compile with fullgraph=True and inference_mode → tangent:", e)


# test case 5: use jvp in torch.compile with fullgraph=False and inference_mode
@torch.compile(backend="eager", fullgraph=False)
def compiled_jvp(x):
    with torch.inference_mode():
        out, tangent = jvp(lambda x: x**2, (x,), (torch.ones(3),))
        return tangent

try:
    tangent_compiled = compiled_jvp(x)
    print("5-use jvp in torch.compile with fullgraph=False and inference_mode → tangent:", tangent_compiled)
except Exception as e:
    print("5-use jvp in torch.compile with fullgraph=False and inference_mode → tangent:", e)
```
1.Before code modification
截屏2026-03-17 21 31 26

2.After code modification
截屏2026-03-17 21 32 42

Comment (Lillian68) [reply torch/_functorch/eager_transforms.py]: As you can see, there must be an underlying issue when using jvp with both torch.compile and inference_mode enabled simultaneously. So I’m still investigating this issue to find a proper fix.

Comment (Lillian68) [reply torch/_functorch/eager_transforms.py]: Hi @soulitzer I’ve fixed the issue in the C++ layer – could you please review the changes? Thank you!
result
